consumer/consumer.go

```diff
@@ -3,177 +3,193 @@ package consumer
-		// it's possible because sarama defines the KError
-		// type as an int16 that supports the error interface.
-		if kerr, ok := err.(sarama.KError); ok && kerr == sarama.ErrConsumerCoordinatorNotAvailable {
+		if err == sarama.ErrConsumerCoordinatorNotAvailable {
 			// We'll add some additional context
-			// information.  This issue comes from the
+			// information. This issue comes from the
 			// RefreshCoordinator call inside sarama, but
 			// it's not reporting the root cause, sadly.
 			//
 			// Even with this information it's rather an
 			// annoying little issue.
 			err = errors.Wrap(err, "__consumer_offsets topic doesn't yet exist, either because no client has yet requested an offset, or because this consumer group is not yet functioning at startup or after rebalancing.")
 		}
-		err = errors.Wrap(err, fmt.Sprintf("failed to create consumer group %q", groupID))
-		return err
-	}
-
-	cgh := consumerGroupHandler{
-		consumer: c,
+		return errors.Wrap(err, fmt.Sprintf("failed to create consumer group %q", groupID))
 	}
-
-	err = c.consumer.Consume(context.Background(), c.handlers.Topics(), cgh)
-	c.Logger.Println("consumer closing")
-	return err
+	c.consumerGroup = cg
+	return nil
 }
 
-// consumerGroupHandler implements sarama.ConsumerGroupHandler interface.
-// ConsumeClaim will be run in a separate goroutine for each topic/partition.
+// consumerGroupHandler implements sarama.ConsumerGroupHandler
+// interface. ConsumeClaim will be run in a separate goroutine for each
+// topic/partition.
 type consumerGroupHandler struct {
 	consumer *Consumer
 }
 
-func (consumerGroupHandler) Setup(_ sarama.ConsumerGroupSession) error   { return nil }
-func (consumerGroupHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
-
-func (c consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
-	logSuffix := fmt.Sprintf(", topic=%q, partition=%d", claim.Topic(), claim.Partition())
-	c.consumer.Logger.Println("partition reading" + logSuffix)
+func (consumerGroupHandler) Setup(sarama.ConsumerGroupSession) error {
+	return nil
+}
 
-	h, _ := c.consumer.handlers.Get(claim.Topic())
+func (consumerGroupHandler) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
 
+func (c consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	ctx := sess.Context()
+	// Note that the get should always return a non-nil topicHandler
+	// because the consumer group won't call ConsumeClaim with
+	// a topic that wasn't explicitly handed to NewConsumerGroup
+	// and it's not possible to remove a topic from the handler list.
+	h := c.consumer.handlers.get(claim.Topic())
 	for msg := range claim.Messages() {
-		m, attempts := c.handleMsg(msg, h)
-		if m == nil {
+		if err := h.handleMessage(ctx, msg, claim); err != nil {
+			if c.consumer.config.Discarded != nil {
+				c.consumer.config.Discarded(msg, err)
+			}
```

Comment:
Doesn't that naming imply that the message will always get "rejected"? How would retries get handled for instance?

Reply:
With the new handler API, the handler itself can handle retries - in my view retrying is more domain-specific and hence closer to kafka-go's domain than felice's. The plan is to add a RetryHandler type (essentially middleware for handlers) that will handle retries like the current logic does, but because it's just an arbitrary wrapper, it becomes easy to tweak the behaviour (for example to retry for a period depending on the message contents).

The default behaviour of kafka-go would be to use a RetryHandler (the API is planned to remain backwardly compatible). This is my current idea as to what the RetryHandler API might look like: ```
// RetryHandler implements Handler by retrying
type RetryHandler struct {
	// Handler holds the handler that will be retried.
	Handler Handler

	// Strategy is used to determine how the handler
	// retries HandleMessage calls. If there are no more available
	// attempts, the last error returned from HandleMessage
	// will be returnefd
	Strategy func(m *Message) retry.Strategy

	// Logf is used to log a message when an attempt
	// to send the message failed.
	Logf func(format string, args ...interface{})

	// IsFatal reports whether the given
	// error should be considered fatal. Fatal
	// errors will cause the retry loop to abort.
	IsFatal func(err error) bool
}

func (h RetryHandler) HandleMessage(ctx context.Context, m *Message, _ PartitionInfo) (int, error)
```

Reply:
Makes sense! :+1:

Reply:
(re: Strategy I would maybe pass both message and error to the strategy func and let it decide which error to return)

Reply:
Yeah, there's bound to be room for changing things there. We'll probably look at some real scenarios and see how it fits. Also, the nice thing about this is it doesn't have to be one-size-fits-all - if you have a special requirement, you can always just roll your own handler or middleware.